### PR TITLE
[SHELL32] Display the version string on About dialog programmatically

### PR DESCRIPTION
**Pull request by Fraizeraust**
[CORE-16571](https://jira.reactos.org/browse/CORE-16571)

### CODE REVIEW

**Comment by Fraizeraust** [comment]
AFTER:
Capture

**Comment by Fraizeraust** [comment]
I suppose one would need to build AMD64 build on MSVC, right?

**Comment by binarymaster** [comment]
@Fraizeraust there is simpler way, see this commit: 726ecddcf2a6f153f1a318989f26e750acce16ac

You can change `/sdk/include/reactos/version.cmake` locally:
```diff
-set(KERNEL_VERSION "${KERNEL_VERSION_MAJOR}.${KERNEL_VERSION_MINOR}.${KERNEL_VERSION_PATCH_LEVEL}")
+set(KERNEL_VERSION "${KERNEL_VERSION_MAJOR}.${KERNEL_VERSION_MINOR}.${KERNEL_VERSION_PATCH_LEVEL}-amd64")
```
^ Just hardcode this cmake variable.

**Comment by Fraizeraust** [comment]
Capture
The "d" letter is almost hidden by the other string. An easier approach that both satisfies the x86 and AMD64 builds is to change the X coordinate of KERNEL_VERSION_STR resource accordingly so that it'll be a little near to the `Version` resource.

**Comment by binarymaster** [comment]
What about not hacking coordinates in layout, but generate full version string in code? :slightly_smiling_face:

**Comment by Fraizeraust** [comment]
The PR is WIP for now. I need to figure out how can I properly concatenate the version strings into one.

**Comment by learn-more** [comment]
your new code block `Append the version string` is already doing close to what you want.

**Comment by Fraizeraust** [comment]
Capture
Capture2
LGTM.

File: dll/win32/shell32/wine/shell32_main.c [hunk 1]
```diff
@@ -1369,6 +1374,9 @@ BOOL WINAPI ShellAboutW( HWND hWnd, LPCWSTR szApp, LPCWSTR szOtherStuff,
     HRSRC hRes;
     DLGTEMPLATE *DlgTemplate;
     BOOL bRet;
+    WCHAR szVersionString[MAX_PATH];
+    WCHAR szKernelVersion[MAX_PATH];
+    WCHAR szKernelVersionBuild[MAX_PATH];
```
Comment (ThFabba) [inline dll/win32/shell32/wine/shell32_main.c]: These are not paths, please use a different define, or simply a number.

File: dll/win32/shell32/wine/shell32_main.c [hunk 2]
```diff
@@ -1381,7 +1389,23 @@ BOOL WINAPI ShellAboutW( HWND hWnd, LPCWSTR szApp, LPCWSTR szOtherStuff,
...
+    */
+    MultiByteToWideChar(CP_ACP, 0, KERNEL_VERSION_STR, -1, szKernelVersion, _countof(szKernelVersion));
+    MultiByteToWideChar(CP_ACP, 0, KERNEL_VERSION_BUILD_STR, -1, szKernelVersionBuild, _countof(szKernelVersionBuild));
+
+    /* Append the version string */
+    StringCchCatW(szVersionString, _countof(szVersionString), szKernelVersion);
+    StringCchCatW(szVersionString, _countof(szVersionString), L" ");
+    StringCchCatW(szVersionString, _countof(szVersionString), L"(");
+    StringCchCatW(szVersionString, _countof(szVersionString), szKernelVersionBuild);
+    StringCchCatW(szVersionString, _countof(szVersionString), L")");
```
Comment (ThFabba) [inline dll/win32/shell32/wine/shell32_main.c]: You could potentially do all this with a single `StringCchPrintfW` call

File: dll/win32/shell32/wine/shell32_main.c [hunk 3]
```diff
@@ -961,6 +964,7 @@ VOID WINAPI Printers_UnregisterWindow(HANDLE hClassPidl, HWND hwnd)
 typedef struct
 {
     LPCWSTR  szApp;
+    LPCWSTR  szOSVersion;
```
Comment (ThFabba) [inline dll/win32/shell32/wine/shell32_main.c]: The location of this file implies that it's synced with Wine. Are we already diverging from upstream here, or should we consider ifdefs or whatever?

Comment (Fraizeraust) [reply dll/win32/shell32/wine/shell32_main.c]: I see Katayama has committed some changes to that file (as per the copyright header) however I'm not sure if he sent such changes to upstream first. For the sake of easier syncing, I'll just wrap ReactOS related coder in an ifdef.

Comment (binarymaster) [reply dll/win32/shell32/wine/shell32_main.c]: > dll/win32/shell32             # Forked at Wine-20071011

According to README.WINE

**Comment by Fraizeraust** [comment]
Comments addressed.

File: dll/win32/shell32/wine/shell32_main.c [hunk 4]
```diff
@@ -1381,7 +1395,22 @@ BOOL WINAPI ShellAboutW( HWND hWnd, LPCWSTR szApp, LPCWSTR szOtherStuff,
     if(!DlgTemplate)
         return FALSE;
 
+#ifdef __REACTOS__
+    /*
+     * KERNEL_VERSION_STR and KERNEL_VERSION_BUILD_STR are ANSI strings so we must convert them to Unicode
+     * (wide characters) before going to concatenate the OS version strings.
+    */
+    MultiByteToWideChar(CP_ACP, 0, KERNEL_VERSION_STR, -1, szKernelVersion, _countof(szKernelVersion));
+    MultiByteToWideChar(CP_ACP, 0, KERNEL_VERSION_BUILD_STR, -1, szKernelVersionBuild, _countof(szKernelVersionBuild));
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shell32_main.c]: Much simpler you can use macros directly where you use them, as show here:
https://git.reactos.org/?p=reactos.git;a=blob;f=base/setup/welcome/welcome.c;hb=e00ea7daddfa21ba692d6d9a22d460d79920c6fc#l1353
https://git.reactos.org/?p=reactos.git;a=blob;f=base/shell/cmd/cmd.c;hb=e00ea7daddfa21ba692d6d9a22d460d79920c6fc#l2011
https://git.reactos.org/?p=reactos.git;a=blob;f=base/shell/cmd/ver.c;hb=e00ea7daddfa21ba692d6d9a22d460d79920c6fc#l138

**Pull request merged**